processing/src/main/java/org/apache/druid/math/expr/ExprEval.java

```diff
@@ -549,10 +551,13 @@ public static ExprEval ofType(@Nullable ExpressionType type, @Nullable Object va
         }
         return ofDouble(null);
       case COMPLEX:
-        // json isn't currently defined in druid-core, this can be reworked once
-        // https://github.com/apache/druid/pull/13698 is merged (or COMPLEX<json> is promoted to a real built-in type(s)
-        if ("json".equals(type.getComplexTypeName())) {
-          return ofComplex(type, value);
+        if (ExpressionType.NESTED_DATA.equals(type)) {
+          // is the thing actually nested?
+          final Object unwrapped = StructuredData.unwrap(value);
+          if (unwrapped instanceof Map) {
+            return ofComplex(type, unwrapped);
+          }
+          return bestEffortOf(unwrapped);
```

Comment:
Seems fishy: someone asked for an `ExprEval.of` a `COMPLEX<json>` but will get some other type. Why's this needed? If it's really necessary can we call it out louder?

Reply:
removed, this was to support mixed type columns which were still not actually nested which a number of the `BaseFilterTest` input data had. I have since changed to present as the least restrictive type instead of always just showing up as `COMPLEX<json>`, so this is no longer necessary. To better support this, I now store these mixed variant types in `VariantArrayColumn` since it already had the basic structure required to support actual variant types and was only being used for single type ARRAY columns previously.